### PR TITLE
Refactor DNN module to build with cudnn 9

### PR DESCRIPTION
**Pull request by ZelboK**
A lot of APIs that are currently being used in the dnn module have been removed in cudnn 9. They were deprecated in 8. This PR updates said code accordingly to the newer API.

Some key notes:
1) This is my first PR. I am new to openCV. 
2) `opencv_test_core` tests pass
3) On a 3080, cuda 12.4(should be irrelevant since I didn't build the `opencv_modules`, gcc 11.4, WSL 2. 
4) For brevity I will avoid including macro code that will allow for older versions of cudnn to build.

I was unable to get the tests working for `opencv_test_dnn` and `opencv_perf_dnn`. The errors I get are of the following: 
```
 OpenCV tests: Can't find required data file: dnn/onnx/conformance/node/test_reduce_prod_default_axes_keepdims_example/model.onnx in function 'findData'
" thrown in the test body.
```
So before I spend more time investigating I was hoping to get a maintainer to point me in the right direction here. I would like to run these tests and confirm things are working as intended. I may have missed some details.

### Pull Request Readiness Checklist

relevant issue
(https://github.com/opencv/opencv/issues/24983

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by ZelboK** [comment]
I ran the precommit hook, but it is highlighting a lot of formatting related problems. Is there a simple fix for this?

**Comment by asmorkalov** [comment]
@ZelboK thanks for the patch. General comments:
- Please  do not touch original formatting and the code that is not relevant to CuDNN 9.0.
- OpenCV uses opencv_extra repository to store test data. See https://github.com/opencv/opencv/wiki/QA_in_OpenCV. Also there is download_models.py script in `opencv_extra/testdata/dnn` to download 3rdparty models for testing. Please use it to achieve full dnn coverage locally.

**Comment by asmorkalov** [comment]
The PR breaks CuDNN 8 (v 8.7.0, Ubuntu  18.04):
<details>

```
[ 72%] Building CXX object modules/dnn/CMakeFiles/opencv_dnn.dir/src/layers/const_layer.cpp.o
In file included from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn.hpp:8:0,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../op_cuda.hpp:12,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/blank_layer.cpp:43:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp: In constructor ‘cv::dnn::cuda4dnn::csl::cudnn::RNNDescriptor<T>::RNNDescriptor(const cv::dnn::cuda4dnn::csl::cudnn::Handle&, cv::dnn::cuda4dnn::csl::cudnn::RNNDescriptor<T>::RNNMode, int, int, bool, const cv::dnn::cuda4dnn::csl::cudnn::DropoutDescriptor&)’:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp:129:11: error: ‘input_size’ was not declared in this scope
           input_size, hidden_size,
           ^
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn/cudnn.hpp:22:53: note: in definition of macro ‘CUDA4DNN_CHECK_CUDNN’
     ::cv::dnn::cuda4dnn::csl::cudnn::detail::check((call), CV_Func, __FILE__, __LINE__)
                                                     ^~~~
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp:129:11: note: suggested alternative: ‘hidden_size’
           input_size, hidden_size,
           ^
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn/cudnn.hpp:22:53: note: in definition of macro ‘CUDA4DNN_CHECK_CUDNN’
     ::cv::dnn::cuda4dnn::csl::cudnn::detail::check((call), CV_Func, __FILE__, __LINE__)
                                                     ^~~~
In file included from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/reshape.hpp:12:0,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/blank_layer.cpp:49:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/tensor_ops.hpp: In constructor ‘cv::dnn::cuda4dnn::csl::LSTM<T>::LSTM(cv::dnn::cuda4dnn::csl::cudnn::Handle, const cv::dnn::cuda4dnn::csl::LSTM<T>::params_type&)’:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/tensor_ops.hpp:548:13: error: ‘filterDesc’ was not declared in this scope
             filterDesc = FilterDescriptor(params.weights_shape);
             ^~~~~~~~~~
modules/dnn/CMakeFiles/opencv_dnn.dir/build.make:853: recipe for target 'modules/dnn/CMakeFiles/opencv_dnn.dir/src/layers/blank_layer.cpp.o' failed
make[3]: *** [modules/dnn/CMakeFiles/opencv_dnn.dir/src/layers/blank_layer.cpp.o] Error 1
make[3]: *** Ожидание завершения заданий…
In file included from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn.hpp:8:0,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../op_cuda.hpp:12,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/const_layer.cpp:10:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp: In constructor ‘cv::dnn::cuda4dnn::csl::cudnn::RNNDescriptor<T>::RNNDescriptor(const cv::dnn::cuda4dnn::csl::cudnn::Handle&, cv::dnn::cuda4dnn::csl::cudnn::RNNDescriptor<T>::RNNMode, int, int, bool, const cv::dnn::cuda4dnn::csl::cudnn::DropoutDescriptor&)’:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp:129:11: error: ‘input_size’ was not declared in this scope
           input_size, hidden_size,
           ^
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn/cudnn.hpp:22:53: note: in definition of macro ‘CUDA4DNN_CHECK_CUDNN’
     ::cv::dnn::cuda4dnn::csl::cudnn::detail::check((call), CV_Func, __FILE__, __LINE__)
                                                     ^~~~
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/cudnn/recurrent.hpp:129:11: note: suggested alternative: ‘hidden_size’
           input_size, hidden_size,
           ^
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/csl/cudnn/cudnn.hpp:22:53: note: in definition of macro ‘CUDA4DNN_CHECK_CUDNN’
     ::cv::dnn::cuda4dnn::csl::cudnn::detail::check((call), CV_Func, __FILE__, __LINE__)
                                                     ^~~~
In file included from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/const.hpp:12:0,
                 from /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/const_layer.cpp:23:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/tensor_ops.hpp: In constructor ‘cv::dnn::cuda4dnn::csl::LSTM<T>::LSTM(cv::dnn::cuda4dnn::csl::cudnn::Handle, const cv::dnn::cuda4dnn::csl::LSTM<T>::params_type&)’:
/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/../cuda4dnn/primitives/../csl/tensor_ops.hpp:548:13: error: ‘filterDesc’ was not declared in this scope
             filterDesc = FilterDescriptor(params.weights_shape);
             ^~~~~~~~~~
modules/dnn/CMakeFiles/opencv_dnn.dir/build.make:881: recipe for target 'modules/dnn/CMakeFiles/opencv_dnn.dir/src/layers/const_layer.cpp.o' failed
```

</details>

**Comment by johnnynunez** [comment]
> The PR breaks CuDNN 8 (v 8.7.0, Ubuntu 18.04):

it was because the PR is incomplete.
Some variables not exists as input_size.
This modules need to refactored

**Comment by johnnynunez** [comment]
great job @asmorkalov! I'm waiting to that for build on jetson

**Comment by asmorkalov** [comment]
There is still LSTM test failure.

File: modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp [hunk 1]
```diff
@@ -528,6 +528,49 @@ namespace cv { namespace dnn { namespace cuda4dnn { namespace csl {
...
+                                    cudnnHandle.get(), rnnDesc.get(), CUDNN_FWD_MODE_INFERENCE,
+                                    xDesc, &workSpaceSize, &reserveSpaceSize));
+
+            std::cout << "workSpaceSize from cudnn: " << workSpaceSize << std::endl;
+            std::cout << "reserveSpaceSize from cudnn: " << reserveSpaceSize << std::endl;
+            csl::WorkspaceBuilder builder;
+            builder.require<T>(workSpaceSize);
+            builder.require<T>(reserveSpaceSize);
+            scratch_mem_in_bytes = builder.required_workspace_size();
+            std::cout << "scratch_mem_in_bytes: " << scratch_mem_in_bytes << std::endl;
```
Comment (asmorkalov) [inline modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp:571]: Let's remove std::cout or conver them to CV_LOG_DEBUG.

File: modules/dnn/src/cuda4dnn/primitives/recurrent_cells.hpp [hunk 2]
```diff
@@ -75,8 +72,8 @@ class LSTMOp final : public CUDABackendNode
         Ptr<wrapper_type> yc_output_wrapper = outputs.size() == 2 ? outputs[1].dynamicCast<wrapper_type>() : Ptr<wrapper_type>();
         csl::TensorSpan<T> yc_output = yc_output_wrapper.empty() ? csl::TensorSpan<T>() : yc_output_wrapper->getSpan();
 
-        csl::WorkspaceAllocator allocator(workspace);
-        lstm.inference(input, y_output, yc_output, filtersTensor, h0Tensor, c0Tensor, allocator.get_instance());
+        lstm.inference(input, y_output, yc_output, filtersTensor, h0Tensor, c0Tensor, workspace);
+        // might need to add reserveSpace here.
```
Comment (asmorkalov) [inline modules/dnn/src/cuda4dnn/primitives/recurrent_cells.hpp:76]: Already done by `builder.require<T>(reserveSpaceSize)`

File: modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp (hunk 1 above)
Comment (ZelboK) [reply modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp]: They can be removed - if I recall correctly I added them for debugging.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**